memory_unsafe.go

```diff
@@ -0,0 +1,256 @@
+	// is collected and reused by the runtime. This has a few reasons:
+	//
+	//  - Avoid leaking memory/mappings.
+	//  - Future writes to this memory should never clobber a bpf map's contents.
+	//  - Some bpf maps are mapped read-only, causing a segfault if the runtime
+	//    reallocates and zeroes the span later.
+	runtime.SetFinalizer((*byte)(addr), unmap(size))
+
+	return nil
+}
+
+// unmap returns a function that takes a pointer to a memory-mapped region on
+// the Go heap. The function undoes any mappings and discards the span's
+// contents.
+//
+// Used as a finalizer in [newMemory], split off into a separate function for
+// testing and to avoid accidentally closing over the unsafe.Pointer to the
+// memory region, which would cause a cyclical reference.
+//
+// The resulting function panics if the mmap operation returns an error, since
+// it would mean the integrity of the Go heap is compromised.
+func unmap(size int) func(*byte) {
+	return func(a *byte) {
+		// Create another mapping at the same address to undo the original mapping.
+		// This will cause the kernel to repair the slab since we're using the same
+		// protection mode and flags as the original mapping for the Go heap.
+		//
+		// Address           Kbytes     RSS   Dirty Mode  Mapping
+		// 000000c000000000    4096     884     884 rw--- [ anon ]
+		//
+		// Using munmap here would leave an unmapped hole in the heap, compromising
+		// its integrity.
+		//
+		// MmapPtr allocates another unsafe.Pointer at the same address. Even though
+		// we discard it here, it may temporarily resurrect the backing array and
+		// delay its collection to the next GC cycle.
+		_, err := unix.MmapPtr(-1, 0, unsafe.Pointer(a), uintptr(size),
+			unix.PROT_READ|unix.PROT_WRITE,
+			unix.MAP_PRIVATE|unix.MAP_FIXED|unix.MAP_ANON)
+		if err != nil {
+			panic(fmt.Errorf("undoing bpf map memory mapping: %w", err))
+		}
+	}
+}
+
+// checkUnsafeMemory ensures value T can be accessed in mm at offset off.
+func checkUnsafeMemory[T any](mm *Memory, off uint64) error {
+	if mm.b == nil {
+		return fmt.Errorf("memory-mapped region is nil")
+	}
+	if mm.ro {
+		return ErrReadOnly
+	}
+	if !mm.heap {
+		return fmt.Errorf("memory region is not heap-mapped, see CollectionOptions.UnsafeVariableExperiment: %w", ErrNotSupported)
+	}
+
+	var t T
+	size := binary.Size(t)
+	if size < 0 {
+		return fmt.Errorf("can't determine size of type %T: %w", t, ErrInvalidType)
+	}
+
+	align := internal.Align(off, uint64(size))
+	if off != align {
+		return fmt.Errorf("unaligned access of memory-mapped region: size %d at offset %d aligns to %d", size, off, align)
+	}
+
+	vs, bs := uint64(size), uint64(len(mm.b))
+	if off+vs > bs {
+		return fmt.Errorf("%d-byte value at offset %d exceeds mmap size of %d bytes", vs, off, bs)
+	}
+
+	return nil
+}
+
+// reinterp reinterprets a pointer of type In to a pointer of type Out.
+func reinterp[Out any, In any](in *In) *Out {
```

Comment:
Nit: just do the cast directly?

Reply:
Sure. The initial proposal had dozens of these, so this made things less error-prone.